### PR TITLE
Dev frontend - home page and login page

### PR DESCRIPTION
**Pull request by michepaolo**
- Created new home page
- Created new Modal to insert a new document description
- Created login page

### CODE REVIEW

File: client/src/API/API.ts [hunk 1]
```diff
@@ -28,9 +30,265 @@ const logout = async () => {
...
+      language: "",
+      pages: null,
+      coordinates: { latitude: 67.86, longitude: 20.231 },
+    },
+  ];
+}
+
+async function postDocument(id: number, title: string, description: string) {
+  try {
+    const response = await fetch(baseURL + `/documents/${id}`, {
```
Comment (umberto-fontanazza) [inline client/src/API/API.ts:175]: post document shouldn't take an id since the id is assigned by the database. The url needs changing, the args of this method need to be edited.

File: client/src/API/API.ts [hunk 2]
```diff
@@ -28,9 +30,265 @@ const logout = async () => {
...
+    });
+    if (response.ok) {
+      return true;
+    }
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+async function patchDocument(id: number, title: string, coordinates: string) {
```
Comment (umberto-fontanazza) [inline client/src/API/API.ts:191]: patch should support a bunch of optional parameters because you may want to edit only some of them. Watch out for nullability this can be done later but leave a //TODO: in the code maybe and fix the type hints

File: client/src/API/API.ts [hunk 3]
```diff
@@ -28,9 +30,265 @@ const logout = async () => {
...
+    });
+    if (response.ok) {
+      return true;
+    }
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+async function deleteDocument(id: number) {
```
Comment (umberto-fontanazza) [inline client/src/API/API.ts:210]: This can be hinted as returning Promise<void> there is no need to return a boolean since the fullfilment of the promise is confirmation of success.

File: client/src/API/API.ts [hunk 4]
```diff
@@ -28,9 +30,265 @@ const logout = async () => {
...
+  try {
+    const response = await fetch(baseURL + `/documents/${id}`, {
+      method: "DELETE",
+      credentials: "include",
+    });
+    if (response.ok) {
+      return true;
```
Comment (umberto-fontanazza) [inline client/src/API/API.ts:217]: Or just return without the bool and raise an exception otherwise.

`if (!response.ok) throw new Error(message)`

File: client/src/API/API.ts [hunk 5]
```diff
@@ -28,9 +30,265 @@ const logout = async () => {
...
+    }
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+/****************************   LINKS   **********************************/
+
+async function getLinks(id: number) {
```
Comment (umberto-fontanazza) [inline client/src/API/API.ts:226]: This route doesn't exist. Check API docs

File: client/src/API/API.ts [hunk 6]
```diff
@@ -28,9 +30,265 @@ const logout = async () => {
...
+    }
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+async function putLink(
+  sourceNodeId: number,
+  targetNodeId: number,
+  linkType: string
```
Comment (umberto-fontanazza) [inline client/src/API/API.ts:241]: better to define an enum for this

File: client/src/API/API.ts [hunk 7]
```diff
@@ -28,9 +30,265 @@ const logout = async () => {
...
+    if (response.ok) {
+      const links = await response.json();
+      return links;
+    }
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+async function putLink(
```
Comment (umberto-fontanazza) [inline client/src/API/API.ts:238]: Check the API documentation for this API route. This is not consistent

File: client/src/App.tsx [hunk 8]
```diff
@@ -1,12 +1,14 @@
...
 import NavHeader from "./components/NavHeader";
 import NotFound from "./components/NotFound";
 import API from "./API/API";
-import "./styles/global.scss"
+import "./styles/global.scss";
 
-function App() {
+const App: FC = () => {
   const [loggedIn, setLoggedIn] = useState(false);
   const [user, setUser] = useState("");
```
Comment (umberto-fontanazza) [inline client/src/App.tsx:13]: I strongly advise for defining a global user context to manage session data. 
You can check an [example](https://github.com/polito-WA1-2024-exam/exam-2-umberto-fontanazza/blob/main/client/src/App.jsx) in this project of mine but this is totally up to you.

File: client/src/components/Home.tsx [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+import ModalAdd from "./ModalForm";
+import MapComponent from "./Map";
+
+interface HomeProps {
+  login: boolean;
+  handleLogout: () => void;
+}
+
+const Home: FC<HomeProps> = (props): JSX.Element => {
+  const [documents, setDocuments] = useState<Document[]>([]);
```
Comment (umberto-fontanazza) [inline client/src/components/Home.tsx:15]: Good job with the type hinting!

File: client/src/components/Home.tsx [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+          <img className="btn-close-img" src="/x.svg" alt="Close" />
+        </button>
+      </div>
+      <div className="content">
+        <img src="/icons8-under-construction-50.png" alt="Under Construction" />
+        <hr />
+        <h3>{props.document?.title}</h3>
+        <p>{props.document?.description}</p>
+        <hr className="separator" />
+        <h6>
```
Comment (umberto-fontanazza) [inline client/src/components/Home.tsx:142]: h6 should be nested inside h4 to have semantic HTML. Use css for font-size instead

File: client/src/components/ModalForm.tsx
Comment (umberto-fontanazza) [inline client/src/components/ModalForm.tsx:1]: This file is very long, consider splitting it

File: client/src/components/NavHeader.tsx [hunk 11]
```diff
@@ -1,18 +1,33 @@
...
+        <div className="nav-left">
+          <img
+            src="/public/LOGO.png"
+            alt="Kiruna eXplorer Logo"
+            className="nav-logo"
+          ></img>
+          <div className="nav-brand">Kiruna eXplorer.</div>
+        </div>
+        {props.login ? (
+          <button className="btn-logout" onClick={props.logout}>
```
Comment (umberto-fontanazza) [inline client/src/components/NavHeader.tsx:23]: Beware dashing create a single class named "btn-logout". You can assign two separate classes intead with className="btn logout" and if you want to select elements with both classes in CSS you just:

`.btn.logout { /* your rules */}`

File: client/src/utils/interfaces.tsx [hunk 12]
```diff
@@ -4,3 +4,17 @@ export interface Credentials {
...
+
+export interface Document {
+  id: number | string;
+  title: string;
+  description: string | null;
+  stakeholder: string | undefined;
+  scale: string | undefined;
+  issuanceDate: Date | null;
+  type: string | undefined;
+  connections: string | undefined;
```
Comment (umberto-fontanazza) [inline client/src/utils/interfaces.tsx:16]: connection is likely a number

File: frontend-notes.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+-   Button "Create new connection" e poi selezione direttamente dalla mappa o anche dal diagramma con doc in risalto
+    
+    VISUALIZZAZIONE CONNECTIONS
+-   
+
+
+
+QUESTIONS
+- Se usiamo la barra laterale -> come lo implementiamo nella visualizzazione da smartphone?
+- Can the stakeholders be more than one?
```
Comment (umberto-fontanazza) [inline frontend-notes.md:35]: Yes

File: client/src/utils/interfaces.tsx (hunk 12 above)
Comment (lorericci) [reply client/src/utils/interfaces.tsx:16]: Ok but this wouldn't match the API draft, where connections is links = [targetDocumentId: number, connectionType: string]. Let connections a number it's easier for me to manage and to show. The correct parameter for the API will be managed in an other way.

Comment (lorericci) [reply client/src/utils/interfaces.tsx:16]: export interface Link {
  targetDocumentId: number;
  type: string[];
}

export interface Document {
  id: number | string;
  title: string;
  description: string | null;
  stakeholder: string | undefined;
  scale: string | undefined;
  issuanceDate: Date | null;
  type: string | undefined;
  connections: Link[] | undefined;
  language: string | undefined;
  pages: number | null;
  coordinates: { latitude: number | null; longitude: number | null };
}

File: client/src/API/API.ts (hunk 1 above)
Comment (lorericci) [reply client/src/API/API.ts:175]: async function postDocument(document: Document){
    try{
       const response = await fetch(baseURL + `/documents`, {...}
     }
}
Document will be something like this: 
export interface Link {
  targetDocumentId: number;
  type: string[];
}

export interface Document {
  id: number | string;
  title: string;
  description: string | null;
  stakeholder: string | undefined;
  scale: string | undefined;
  issuanceDate: Date | null;
  type: string | undefined;
  connections: Link[] | undefined;
  language: string | undefined;
  pages: number | null;
  coordinates: { latitude: number | null; longitude: number | null };
} 
and in the body i will pass all the parameters.

File: client/src/API/API.ts (hunk 2 above)
Comment (lorericci) [reply client/src/API/API.ts:191]: For now this API should be use for the connections, indeed if i want to add a new connection to a document i will call this API.

File: client/src/API/API.ts (hunk 3 above)
Comment (lorericci) [reply client/src/API/API.ts:210]: You mean something like this?
async function deleteDocument(id: number) {
  try {
    const response = await fetch(baseURL + `/documents/${id}`, {
      method: "DELETE",
      credentials: "include",
    });
  } catch (err) {
    console.error(err);
  }
}

File: client/src/API/API.ts (hunk 4 above)
Comment (lorericci) [reply client/src/API/API.ts:217]: Ok i'll do this i prefer.

File: client/src/API/API.ts (hunk 5 above)
Comment (lorericci) [reply client/src/API/API.ts:226]: modified

File: client/src/API/API.ts (hunk 7 above)
Comment (lorericci) [reply client/src/API/API.ts:238]: async function putLink(targetDocumentId: number, linkType: string, id: number) {
  try {
    const response = await fetch(baseURL + `/${id}/links`, {
      method: "PUT",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ targetDocumentId, linkType }),
    });
    if (response.ok) {
      return true;
    }
  } catch (err) {
    console.error(err);
  }
}

File: client/src/API/API.ts (hunk 6 above)
Comment (lorericci) [reply client/src/API/API.ts:241]: Not useful anymore

File: client/src/components/Home.tsx (hunk 9 above)
Comment (lorericci) [reply client/src/components/Home.tsx:15]: I know :)

File: client/src/components/Home.tsx (hunk 10 above)
Comment (lorericci) [reply client/src/components/Home.tsx:142]: Ok, makes sense

File: client/src/API/API.ts (hunk 1 above)
Comment (umberto-fontanazza) [reply client/src/API/API.ts:175]: document id is a number. Also description is mandatory. We need a separate discussion on connections

File: client/src/API/API.ts (hunk 3 above)
Comment (umberto-fontanazza) [reply client/src/API/API.ts:210]: Yes but also check the response status
`if (!response.ok) throw new Error(msg)`. I am not sure about the exceptions you could get with the fetch

Comment (umberto-fontanazza) [reply client/src/API/API.ts:210]: Yes but with the response status check

File: client/src/API/API.ts (hunk 7 above)
Comment (umberto-fontanazza) [reply client/src/API/API.ts:238]: If you read carefully the API docs for the route PUT /documents/{id}/links the array of the link object you pass through the request body replaces the previous array.

TLDR; `linkTypes: string[]`

File: client/src/utils/interfaces.tsx (hunk 12 above)
Comment (umberto-fontanazza) [reply client/src/utils/interfaces.tsx:16]: Most likely this was just a misunderstanding because I called link the complete representation of what happens between two nodes and connections the count of outgoing links from a node

File: client/src/API/API.ts (hunk 2 above)
Comment (umberto-fontanazza) [reply client/src/API/API.ts:191]: to add/edit links you should use PUT /documents/{id}/links. 
Patch /documents/{id} is not used for links (connections)

**Review by umberto-fontanazza — APPROVED** [review]
Approved the merge after I saw those nice SCSS variables
